TickSpec/TickSpec.fs

```diff
@@ -98,69 +111,103 @@ type StepDefinitions (givens,whens,thens,events,valueParsers) =
         StepDefinitions(assembly.GetTypes())
     /// Constructs instance by reflecting against specified types
     new (types:Type[]) =
-        let getScope attributes =
-            attributes
-            |> Seq.cast
-            |> Seq.fold (fun (tags,features,scenarios) (x:StepScopeAttribute) ->
-                x.Tag::tags, x.Feature::features, x.Scenario::scenarios
-            ) ([],[],[])
         let methods =
             types
-            |> Seq.collect (fun t ->
-                let attributes = t.GetCustomAttributes(typeof<StepScopeAttribute>,true)
-                let tags, features, scenarios = getScope attributes
-                t.GetMethods()
-                |> Seq.map (fun m ->
-                    let attributes = m.GetCustomAttributes(typeof<StepScopeAttribute>,true)
-                    let tags', features', scenarios' = getScope attributes
-                    tags@tags' |> List.filter (not << String.IsNullOrEmpty),
-                        features@features' |> List.filter (not << String.IsNullOrEmpty),
-                            scenarios@scenarios' |> List.filter (not << String.IsNullOrEmpty),
-                                m
+            |> Seq.collect (fun t -> t.GetMethods())
+        StepDefinitions(methods)
+    internal new (methods:MethodInfo seq) =
+        let categorizedMethods =
+            let getScope attributes =
```

Comment:
much more efficient to do it this way
probably also the add calls should have the filtering at this point too
```
          let getScope attributes =
                let t, f, s = ResizeArray(), ResizeArray(), ResizeArray()
                let add (x: ResizeArray) s = if not (String.isNullOrEmpty s) then x.Add s
                for x in attributes }> Seq.cast<StepScopeAttribute> do
                    add t x.Tag
                    add f x.Feature
                    add s x.Scenario
                t.ToArray(), f.ToArray(). s.ToArray()
```                
if the reversed order is necessary, `x.Add` can become `x.Insert(0,x)`

Reply:
Considering that I don't care about efficiency so much here (there will be only a few values), I decided to leave the implementation similar with incorporated suggestions from the other comment. I also added the `add` method as you suggested, that is a great point. No reason to filter it afterwards.

Reply:
👍 It all reads nice and clear now, and that's the main thing...